### PR TITLE
Improve flexibility and correctness

### PR DESCRIPTION
**Pull request by erszcz**
Be more flexible in terms of extensions, but more strict with regard to correctness of use:

- allow for specifying `connection_steps` from the user spec

- require passing connection module as is, not a shorthand form (i.e. `escalus_tcp` not just `tcp`) - such modules can now be developed outside Escalus and treated equally to the provided implementations

- complain when an atom is passed as a stanza Recipient, as it's not clear in such a case whether the atom should be translated to a bare or a full JID - _explicit is better than implicit_

- escalus_tcp: allow to override `parser_opts` from the user spec

None of the changes is breaking current API, but some new warnings will be generated. Old code is marked with `TODO`s for future removal.

### CODE REVIEW

File: src/escalus_connection.erl [hunk 1]
```diff
@@ -216,12 +215,45 @@ maybe_forward_to_owner(_, State, Stanzas, Fun) ->
...
+-spec get_module(module() | atom()) -> ?MODULE:t().
+get_module(M) when is_atom(M) ->
+    case is_deprecated_connection_module(M) of
+        true ->
+            %% This function is a natural extension point - we could extend Escalus freely
+            %% by providing modules that implement escalus_connection behaviour.
+            %% However, due to this mapping (predefined atom -> module name)
+            %% we limit this flexibility.
+            %% TODO: Let's remove this limitation in the future.
+            Msg = io_lib:format("~s:get_module/1 called with transport '~s' (use a module name instead)",
```
Comment (gadgetci) [inline src/escalus_connection.erl]: According to **Elvis**:
> Line 236 is too long:             Msg = io_lib:format("~s:get_module/1 called with transport '~s' (use a module name instead)",.

File: src/escalus_pred.erl [hunk 2]
```diff
@@ -642,6 +643,12 @@ is_compressed(#xmlel{name = <<"compressed">>} = Stanza) ->
 is_compressed(_) ->
     false.
 
+is_bind_result(#xmlel{} = Stanza) ->
```
Comment (gadgetci) [inline src/escalus_pred.erl]: According to **Compiler**:
> Warning: missing specification for function is_bind_result/1

File: src/escalus_stanza.erl [hunk 3]
```diff
@@ -277,7 +277,19 @@ error_element(Type, Condition) ->
            children = [#xmlel{name = Condition,
                               attrs = [{<<"xmlns">>, ?NS_STANZA_ERRORS}]}]}.
 
+message(From, Recipient, Type, Msg) when is_atom(Recipient) ->
```
Comment (gadgetci) [inline src/escalus_stanza.erl]: According to **Compiler**:
> Warning: missing specification for function message/4

File: src/escalus_connection.erl [hunk 4]
```diff
@@ -137,7 +136,7 @@ prepare_step(Fun) when is_function(Fun, 3) ->
     Fun.
 
 connect(Props) ->
```
Comment (gadgetci) [inline src/escalus_connection.erl]: According to **Compiler**:
> Warning: missing specification for function connect/1

File: src/escalus_users.erl [hunk 5]
```diff
@@ -243,8 +243,8 @@ get_user_by_name(Name) ->
 
 -spec create_user(escalus:config(), named_user()) -> any().
 create_user(Config, {_Name, Options}) ->
-    ClientProps = get_options(Config, Options),
-    {ok, Conn, ClientProps, _} = escalus_connection:start(ClientProps,
+    ClientProps0 = get_options(Config, Options),
+    {ok, Conn, ClientProps, _} = escalus_connection:start(ClientProps0,
```
Comment (gadgetci) [inline src/escalus_users.erl]: According to **Elvis**:
> The code in the following (LINE, COL) locations has the same structure: (247, 32), (302, 22).

Comment (erszcz) [inline src/escalus_users.erl]: Won't fix. This is intentional - the line in question is a single function call. The equivalence of passed in arguments is coincidental.

File: src/escalus_stanza.erl [hunk 6]
```diff
@@ -277,7 +277,24 @@ error_element(Type, Condition) ->
            children = [#xmlel{name = Condition,
                               attrs = [{<<"xmlns">>, ?NS_STANZA_ERRORS}]}]}.
 
+-spec message(From, Recipient, Type, Msg) -> #xmlel{} when
```
Comment (gadgetci) [inline src/escalus_stanza.erl]: According to **Elvis**:
> The spec in line 280 uses a record, please define a type for the record and use that instead.

**Pull request merged**